packages/amplify-cli/src/version-gating/index.ts

```diff
@@ -0,0 +1,202 @@
+
+  /*
+
+  Summary for the version gating logic:
+
+  Inputs:
+  - M: Metadata in root stack
+  - CV: Currently running CLI verison
+  - CMin: Minimum version value defined in currently running CLI's package.json
+  - DV: Version of the CLI the root stack was deployed with last time
+  - DMin: Minimum required CLI version for deployment
+
+  Checks:
+  - M == null or no version info present => pass (Current CLI is newer always)
+  - CV >= DMin => pass (CV >= CMin always so no need to check)
+  - In all other cases => fail (for example: DMin > CV)
+
+  Notes:
+  - DV: It is possible that this information will change from 5.1.0 to 5.0.2 and 5.1.3 with pushes as long as DMin is 5.0.0.
+  - DV: is not used for any checks, just persisted for diagnostic purposes.
+
+  Cases we cannot handle:
+  - If a CLI version is used which does not have version gating, will always can run against a newer stack, which will nuke out
+    the metadata we added.
+
+  */
+
+  const cfnClient: CloudFormation = cloudFormation.cfn as CloudFormation;
+
+  const templateSummary = await cfnClient
+    .getTemplateSummary({
+      StackName: rootStackName,
+    })
+    .promise();
+
+  const metadataValue = _.get(templateSummary, ['Metadata']) || '{}';
+  const metadata = JSONUtilities.parse(metadataValue);
+  const versionGatingMetadata: VersionGatingMetadata = _.get(metadata, ['AmplifyCLI']);
+
+  // If no metadata or version info in metadata was found in the root stack template, then
+  // assume version check passed as current CLI is newer than the stack was deployed with.
+  if (
+    versionGatingMetadata === undefined ||
+    versionGatingMetadata.MinimumCompatibleCLIVersion === undefined ||
+    versionGatingMetadata.DeployedByCLIVersion === undefined
+  ) {
+    return true;
+  }
+
+  // These are always valid version numbers as CLI creating them.
+  const stackMinimumCompatibleCLIVersion = semver.coerce(versionGatingMetadata.MinimumCompatibleCLIVersion)!;
+
+  // Pick the greater minimum version
+  const minimumCompatibleCLIVersion = semver.gt(stackMinimumCompatibleCLIVersion, context.versionInfo.minimumCLIVersion)
+    ? stackMinimumCompatibleCLIVersion
+    : semver.coerce(context.versionInfo.minimumCLIVersion)!;
+
+  // If current version is greater than
+  if (semver.gte(context.versionInfo.currentCLIVersion, minimumCompatibleCLIVersion)) {
+    return true;
+  }
+
+  const deployedCLIVersion = semver.coerce(versionGatingMetadata.DeployedByCLIVersion)!;
+
+  context.print.warning(
+    `This project was previously deployed with Amplify CLI version ${deployedCLIVersion}. The currently running Amplify CLI version is ${context.versionInfo.currentCLIVersion}.`,
+  );
+
+  context.print.info('');
+  context.print.info(`Some features in this project require Amplify CLI version >=${minimumCompatibleCLIVersion} to function correctly.`);
+  context.print.info('Upgrade to the latest version of Amplify CLI, run: "amplify upgrade" or "npm install -g @aws-amplify/cli"');
```

Comment:
Nit: could use `isPackaged` from cli-core to conditionally print the pkg or npm upgrade instructions

Reply:
Came from @renebrandel leaving it as it is for now